stan/math/prim/mat/fun/ocl_gpu.hpp

```diff
@@ -0,0 +1,296 @@
+    if (initialized_ == 0) {
+      init();
+      initialized_ = 1;
+    }
+    return oclQueue_;
+  }
+  inline int maxWorkgroupSize() { return max_workgroup_size; }
+};
+
+static ocl ocl_context_queue;
+
+/**
+ * Returns the description of the OpenCL
+ * platform and device that is used.
+ *
+ */
+inline std::string get_description() { return ocl_context_queue.description(); }
+
+/**
+ * Returns the reference to the
+ * OpenCL context. If no context was created,
+ * a new context is created.
+ *
+ */
+inline cl::Context &get_context() { return ocl_context_queue.context(); }
+/**
+ * Returns the reference to the active
+ * OpenCL command queue. If no context
+ * and queue were created,
+ * a new context and queue are created and
+ * the reference to the new queue is returned.
+ *
+ */
+inline cl::CommandQueue &get_queue() { return ocl_context_queue.queue(); }
+/**
+ * Returns the reference to the active
+ * OpenCL command queue. If no context
+ * and queue were created,
+ * a new context and queue are created and
+ * the reference to the new queue is returned.
+ *
+ */
+inline int get_maximum_workgroup_size() {
+  return ocl_context_queue.maxWorkgroupSize();
+}
+
+/**
+ * Compiles all the kernel in the specified group
+ *
+ * @param group The kernel group name
+ *
+ */
+inline void compile_kernel_group(std::string group) {
+  cl::Context &ctx = get_context();
+  std::vector<cl::Device> devices = ctx.getInfo<CL_CONTEXT_DEVICES>();
+  std::string kernel_source = kernel_strings[group];
+  cl::Program::Sources source(
+      1, std::make_pair(kernel_source.c_str(), kernel_source.size()));
+  cl::Program program_ = cl::Program(ctx, source);
+  try {
+    char temp[100];
+    int local = 32;
+    int gpu_local_max = sqrt(get_maximum_workgroup_size());
+    if (gpu_local_max < local)
+      local = gpu_local_max;
+    // parameters that have special limits are for now handled here
+    // kernels with paramters will be compiled separately
+    // for now we have static parameters, so this will be OK
+    snprintf(temp, sizeof(temp), "-D TS=%d -D TS1=%d -D TS2=%d ", local, local,
+             local);
+    program_.build(devices, temp);
+
+    cl_int err = CL_SUCCESS;
+    // Iterate over the kernel list
+    // and get all the kernels from this group
+    for (std::map<std::string, std::string>::iterator it
+         = kernel_groups.begin();
+         it != kernel_groups.end(); ++it) {
+      if (group.compare((it->second).c_str()) == 0) {
+        kernels[(it->first).c_str()]
+            = cl::Kernel(program_, (it->first).c_str(), &err);
+      }
+    }
+  } catch (const cl::Error &e) {
+    domain_error(
+        "OpenCL Initialization", e.what(), e.err(), "\nRetrieving build log\n",
+        program_.getBuildInfo<CL_PROGRAM_BUILD_LOG>(devices[0]).c_str());
+  }
+}
+
+/**
+ * Returns the reference to the compiled kernel.
+ * If the kernel has not yet been compiled,
+ * the kernel group is compiled first.
+ *
+ * @param name The kernel name
```

Comment:
include side-effects in the doc. Looks like this function touches at least:
- `compiled_kernels`
- `kernels`

anything else?